src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1112,6 +1112,109 @@ protected static ImmutableArray<string> CheckTupleElementNames(int cardinality,
+            ITypeSymbol throughType = null)
+        {
+            if (symbol is null)
+            {
+                throw new ArgumentNullException(nameof(symbol));
+            }
+
+            if (within is null)
+            {
+                throw new ArgumentNullException(nameof(within));
+            }
+
+            if (!(within is INamedTypeSymbol || within is IAssemblySymbol))
+            {
+                throw new ArgumentException(CodeAnalysisResources.IsSymbolAccessibleBadWithin, nameof(within));
+            }
+
+            checkInCompilation(symbol, nameof(symbol));
+            checkInCompilation(within, nameof(within));
+            if (!(throughType is null))
+            {
+                checkInCompilation(throughType, nameof(throughType));
+            }
+
+            return IsSymbolAccessibleWithinCore(symbol, within, throughType);
+
+            void checkInCompilation(ISymbol s, string parameterName)
+            {
+                var containingAssembly = computeContainingAssembly(s);
+                if (!assemblyIsOk(containingAssembly))
+                {
+                    throw new ArgumentException(string.Format(CodeAnalysisResources.IsSymbolAccessibleWrongAssembly, parameterName), parameterName);
+                }
+            }
+
+            bool assemblyIsOk(IAssemblySymbol a)
+            {
+                if (a == this.Assembly)
+                {
+                    return true;
+                }
+
+                foreach (var reference in this.References)
+                {
+                    var moduleOrAssembly = this.CommonGetAssemblyOrModuleSymbol(reference);
+                    var assembly = moduleOrAssembly as IAssemblySymbol ?? moduleOrAssembly.ContainingAssembly;
+                    if (a == assembly)
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            IAssemblySymbol computeContainingAssembly(ISymbol s)
+            {
+                while (true)
+                {
+                    switch (s.Kind)
+                    {
+                        case SymbolKind.Assembly:
+                            return (IAssemblySymbol)s;
+                        case SymbolKind.PointerType:
+                            s = ((IPointerTypeSymbol)s).PointedAtType;
+                            continue;
+                        case SymbolKind.ArrayType:
+                            s = ((IArrayTypeSymbol)s).ElementType;
+                            continue;
+                        case SymbolKind.Alias:
+                            s = ((IAliasSymbol)s).Target;
+                            continue;
+                        case SymbolKind.Discard:
+                            s = ((IDiscardSymbol)s).Type;
+                            continue;
+                        case SymbolKind.DynamicType:
+                        case SymbolKind.ErrorType:
```

Comment:
>case SymbolKind.ErrorType: [](start = 24, length = 26)

Some error type symbols actually know what assembly they belong to. I think we should check ContainingAssembly property, and only if it is null default to the compilation's assembly.